alf/algorithms/off_policy_algorithm.py

```diff
@@ -229,6 +229,39 @@ def _update(self, experience, weight):
 
         info = alf.nest.utils.stack_nests(info_list)
         info = dist_utils.params_to_distributions(info, self.train_info_spec)
+        return info
+
+    def _collect_train_info_parallelly(self, experience):
+        batch_size = experience.step_type.shape[1]
+        length = experience.step_type.shape[0]
+
+        exp = alf.nest.map_structure(lambda x: x.reshape(-1, *x.shape[2:]),
+                                     experience)
+
+        if self._use_rollout_state:
+            policy_state = exp.state
+        else:
+            policy_state = self.get_initial_train_state(exp.step_type.shape[0])
+
+        exp = dist_utils.params_to_distributions(
+            exp, self.processed_experience_spec)
+        policy_step = self.train_step(exp, policy_state)
+
+        if self._train_info_spec is None:
+            self._train_info_spec = dist_utils.extract_spec(policy_step.info)
+        info = dist_utils.distributions_to_params(policy_step.info)
+        info = alf.nest.map_structure(
+            lambda x: x.reshape(length, batch_size, *x.shape[1:]), info)
+        info = dist_utils.params_to_distributions(info, self.train_info_spec)
+        return info
+
+    def _update(self, experience, weight):
+        length = experience.step_type.shape[0]
+        if self._config.temporally_independent_train_step or length == 1:
```

Comment:
Do you think it's possible to automatically decide which one to choose depending on other arguments?

Reply:
I thought about it. But feel there is no reliable of doing this. For example SarsaAlgorithm has non-empty train_state even if RNN is not used, but we can still use this option. But in general, this option should not be used for algorithms with states.